REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [2]:
import pandas as pd

In [3]:
bus = pd.read_csv('businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('inspections.csv')

In [4]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [5]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [11]:
records_num = len(bus)
print(f'There are {records_num-1} records')

There are 6314 records


#### Question 2b: How many unique business IDs are there?  

In [56]:
bus['business_id'].unique()

array([   10,    19,    24, ..., 90237, 90238, 90239], dtype=int64)

In [15]:
bus['business_id'].describe()

<bound method NDFrame.describe of 0          10
1          19
2          24
3          31
4          45
        ...  
6310    90234
6311    90236
6312    90237
6313    90238
6314    90239
Name: business_id, Length: 6315, dtype: int64>

In [21]:
bus.iloc[0:10, 0:2]

business_id                                name
0           10                    TIRAMISU KITCHEN
1           19               NRGIZE LIFESTYLE CAFE
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY
3           31      NORMAN'S ICE CREAM AND FREEZES
4           45                 CHARLIE'S DELI CAFE
5           48                          ART'S CAFE
6           50                          SUSHI ZONE
7           54                 RHODA GOLDMAN PLAZA
8           56                          CAFE X + O
9           58                         OASIS GRILL

In [26]:
unique_id = bus['business_id'].unique()
len(unique_id)

6315

In [205]:
bus_count = bus['business_id'].count()
print(f'There are {bus_count} unique business ids')

There are 6315 unique business ids


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [34]:
unique_names = bus['name'].unique()

In [95]:
bus['name'].count()

6315

In [37]:
bus['name'].value_counts()

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
                              ..
7-MISSION RESTAURANT           1
WONDERFUL DESSERT & CAFE       1
KITCHEN ISTANBAL               1
FAT BELLI DELI                 1
ALBA RAY'S                     1
Name: name, Length: 5692, dtype: int64

In [99]:
sf_bus = bus[bus['city']=='San Francisco']
sf_bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [101]:
len(sf_bus)

6315

In [173]:
bus_counts = sf_bus['name'].value_counts()

In [204]:
bus_counts.head(5)

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [20]:
bus.dtypes

business_id       int64
name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
phone_number     object
dtype: object

#### Question 3b

What are the unique values of postal_code?

In [57]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [58]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [70]:
bus_valid = bus[bus['postal_code'].isin(validZip)]

In [71]:
bus_valid.describe()

business_id     latitude    longitude
count   5999.000000  3516.000000  3516.000000
mean   47485.367061    37.773820  -122.425356
std    33977.547969     0.022924     0.027374
min       10.000000    37.668824  -122.510896
25%     5798.000000    37.760526  -122.436389
50%    63943.000000    37.781118  -122.418515
75%    78780.000000    37.790124  -122.406193
max    90239.000000    37.808931  -122.376657

In [73]:
bus_valid = bus[bus['postal_code'].isin(validZip)]

In [74]:
len(bus_valid)

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [79]:
vb_null = vb[pd.isnull(vb['longitude'])]

In [82]:
len(vb_null)

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [189]:
valid_zip_tab = pd.DataFrame(columns=['postal_code', 'null_lon', 'not_null_lon'])


for target_zip in validZip:
    valid_zips = bus[bus['postal_code'] == target_zip]
    
    null_count = valid_zips['longitude'].isnull().sum()
    not_null_count = valid_zips['longitude'].notnull().sum()
    
    valid_zip_tab = valid_zip_tab.append({'postal_code': int(target_zip),
                                          'null_lon': null_count,
                                          'not_null_lon': not_null_count},
                                         ignore_index=True)

valid_zip_tab.head()

C:\Users\nmcco\AppData\Local\Temp\ipykernel_15320\432894991.py:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  valid_zip_tab = valid_zip_tab.append({'postal_code': int(target_zip),
C:\Users\nmcco\AppData\Local\Temp\ipykernel_15320\432894991.py:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  valid_zip_tab = valid_zip_tab.append({'postal_code': int(target_zip),
C:\Users\nmcco\AppData\Local\Temp\ipykernel_15320\432894991.py:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  valid_zip_tab = valid_zip_tab.append({'postal_code': int(target_zip),
C:\Users\nmcco\AppData\Local\Temp\ipykernel_15320\432894991.py:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.

postal_code null_lon not_null_lon
0       94102      196          262
1       94103      252          306
2       94104       59           74
3       94105      100          126
4       94107      247          204

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [192]:
valid_zip_tab['total_zips'] = valid_zip_tab['null_lon'] + valid_zip_tab['not_null_lon']
valid_zip_tab.head()

postal_code null_lon not_null_lon total_zips
0       94102      196          262        458
1       94103      252          306        558
2       94104       59           74        133
3       94105      100          126        226
4       94107      247          204        451

In [203]:
valid_zip_tab['null_lon_%'] = valid_zip_tab['null_lon']/valid_zip_tab['total_zips']
vzts = valid_zip_tab.sort_values(['null_lon_%'], ascending=False)
vzts.head()
print(vzts['postal_code'].head())

4     94107
18    94124
21    94132
10    94114
16    94122
Name: postal_code, dtype: object


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [107]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

In [106]:
len(ins)

15430

In [109]:
unique_ins = ins['business_id'].unique()

In [110]:
len(unique_ins)

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [132]:
ins2 = ins.copy()

In [133]:
ins2['type'].unique()

array(['routine', 'complaint'], dtype=object)

In [134]:
ins2_type_counts = ins2['type'].value_counts()

In [135]:
print(ins2_type_counts)

routine      15429
complaint        1
Name: type, dtype: int64


In [136]:
ins2 = ins2[ins2['type'] == 'routine']

In [137]:
ins2['type'].value_counts()

routine    15429
Name: type, dtype: int64

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [138]:
ins2['date'] = pd.to_datetime(ins2['date'], format='%Y%m%d')

In [139]:
ins2.head()

business_id  score       date     type
0           10     82 2016-05-03  routine
1           10     94 2014-07-29  routine
2           10     92 2014-01-14  routine
3           19     94 2016-05-13  routine
4           19     94 2014-11-10  routine

In [140]:
ins2['year'] = ins2['date'].dt.year

In [141]:
ins2.head()

business_id  score       date     type  year
0           10     82 2016-05-03  routine  2016
1           10     94 2014-07-29  routine  2014
2           10     92 2014-01-14  routine  2014
3           19     94 2016-05-13  routine  2016
4           19     94 2014-11-10  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [151]:
year_min = min(ins2['year'])
year_max = max(ins2['year'])
print(f'the range of years covered by the data set is {year_min} - {year_max}')

the range of years covered by the data set is 2013 - 2016


In [154]:
print(f"the range of years covered by the data set is {min(ins2['year'])} - {max(ins2['year'])}")

the range of years covered by the data set is 2013 - 2016


In [155]:
ins2['year'].value_counts()

2016    5839
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [165]:
ins3 = ins2[ins2['year'].isin([2014,2015,2016])]
ins3['year'].value_counts()

2016    5839
2014    5629
2015    3923
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [167]:
ins_2016 = ins3[ins3['year']==2016]
ins_2016['year'].value_counts()

2016    5839
Name: year, dtype: int64

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [172]:
bus_ins = pd.merge(bus, ins_2016, on='business_id', how='outer')
bus_ins.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score       date     type    year  
0 -122.403816  +14154217044   82.0 2016-05-03  routine  2016.0  
1 -122.421547  +14157763262   94.0 2016-05-13  routine  2016.0  
2 -122.403135  +14156779494   98.0 2016-10-05  routine  2016.0  
3 -122.403135  +14156779494   96.0 2016-03-11  routine  2016.0  
4 -122.419004           NaN   84.0 2016-08-16  routine  2016.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [186]:
ins_score_ranked = bus_ins.sort_values(by='score')
bottom_20 = ins_score_ranked.head(20)

In [187]:
for index, row in bottom_20.iterrows():
    print(f"Name: {row['name']}")
    print(f"Address: {row['address']}")
    print(f"Rating: {row['score']}")
    print("-------------------------")

Name: GOLDEN RIVER RESTAURANT
Address: 5827 GEARY BLVD 
Rating: 52.0
-------------------------
Name: POKI TIME
Address: 2101 LOMBARD ST 
Rating: 55.0
-------------------------
Name: CRAZY PEPPER
Address: 2257 SAN JOSE AVE 
Rating: 55.0
-------------------------
Name: UNCLE CAFE
Address: 65 WAVERLY PL 
Rating: 55.0
-------------------------
Name: HING WANG BAKERY
Address: 339 JUDAH ST 
Rating: 56.0
-------------------------
Name: GOLDEN WOK
Address: 295 B ORIZABA AVE 
Rating: 56.0
-------------------------
Name: TOP'S CAFE
Address: 2460 SAN BRUNO AVE 
Rating: 57.0
-------------------------
Name: L & G VIETNAMESE SANDWICH
Address: 602 EDDY ST 
Rating: 57.0
-------------------------
Name: RED A BAKERY
Address: 634 CLEMENT ST 
Rating: 57.0
-------------------------
Name: SAN TUNG RESTAURANT LLC
Address: 1031 IRVING ST 
Rating: 57.0
-------------------------
Name: MA'S DIM SUM & CAFE
Address: 1315 POWELL ST 
Rating: 57.0
-------------------------
Name: NEW GARDEN RESTAURANT, INC.
Address: 7

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!